### ISSUE TITLE
Access to ScriptServer

### ISSUE THREAD

**Issue by Kehom** [godotengine/godot-cpp#1516]
### Godot version

4.3 beta 2

### godot-cpp version

master

### System information

Windows 10 and Linux Mint

### Issue description

If implementing a module, we can include the `script_language.h` file then access `ScriptServer`. However this is not exposed to GDExtension. I don't know if there is a reason for that.

That said, there are a few things that are very useful. Of particular interest to me is querying if a scripted class inherits from a specific base. I know that we can manually do that by first obtaining the list of "global classes" (`ProjectSettings::get_singleton()->get_global_class_list()`. However this option means that we always have to iterate through the list of classes. Caching the data might result in outdated data. From my understanding `ScriptServer` is always updated whenever a scripted (global) class is created or removed.

I understand if not exposing `ScriptServer` is actually desired.

### Steps to reproduce

-

### Minimal reproduction project

-

**Comment by AThousandShips** [comment]
Feature proposals are tracked [here](https://github.com/godotengine/godot-proposals), this is for bugs. Please open a proposal instead.

**Comment by Kehom** [comment]
Err.. Quite frankly I didn't know if it was exactly a bug or a feature proposal because we can include `script_language.hpp`, albeit the class in there only contains the `ScriptNameCasing` enum.

That said, sorry if this was indeed meant to be a feature proposal.

**Comment by AThousandShips** [comment]
I think it's intentional to give access to the relevant code, but let's see what the feedback is, my bad!

**Comment by dsnopek** [comment]
Whether or not to expose `ScriptServer` the usual way is more a question for whoever is responsible for the Scripting API, rather than something for the GDExtension team to decide.

However, it does appear as though `ScriptServer` is not exposed the normal way on purpose: if it were, then GDScript (or other scripting languages) could interact with it, which seems undesirable.

So, assuming we don't want to expose it in the normal way, we _could_ add some new functions to `gdextension_interface.h` in order to access it, and then in godot-cpp make a wrapper class that mimicks the API of `ScriptServer`?

**Comment by Kehom** [comment]
I do believe `ScriptServer` should not be exposed to GDScript. That said, since the intention of GDExtension is to have parity with Modules, then I do believe we need some sort of access to that server. So indeed, not exactly the "normal way".

**Comment by aloiscochard** [comment]
It would be very useful to be able to use `ScriptServer` from a `GDExtension`, for example it would allow something like [luascript](https://github.com/perbone/luascript) to be installable as an add-on from the asset store as opposed to having to build the engine from source.

That's a pretty big deal in my humble opinion.

**Comment by dsnopek** [comment]
@aloiscochard It's definitely possible to add scripting languages to Godot via a GDExtension, that doesn't require access to `ScriptServer`. I'm not sure why the luascript project you linked to is implemented as a module rather than a GDExtension - if there are technical limitations they encountered, I'd be very curious to learn what they are!

**Comment by aloiscochard** [comment]
@dsnopek I don't know for luascript project as I was really using it as an example to integrate my own scripting language.

I thought I need to access ScriptServer to do those operations: https://github.com/perbone/luascript/blob/master/register_types.cpp#L55-L56

How to register a new file extension and it's handler without access to ScriptServer? If you know of a project doing this already I'll be very happy to dive in there and find the answer myself.

Thanks for your help!

**Comment by dsnopek** [comment]
@aloiscochard 

> How to register a new file extension and it's handler without access to ScriptServer? If you know of a project doing this already I'll be very happy to dive in there and find the answer myself.

This is a test project adding Gravity script support that I use for testing the ability to registry scripting languages, but it's very incomplete:

https://gitlab.com/snopek-games/godot-gravity-lang

And the Luau GDExtension is a full extension that adds support for the Luau scripting language:

https://github.com/Fumohouse/godot-luau-script

**Comment by CedNaru** [comment]
I don't think that ScriptServer should be accessible as it is. But I want to point out that GDScript is still making some uses of it, mostly to access global names, which allows it to use scripts made in any other languages. I could see this being useful for other dynamically typed languages too. Currently, I have no idea how I would access global names using a GDExtension instead of a module (outside the global names defined by the language implementation itself) but maybe it exists, and I didn't find the the api call for it.

**Comment by Kehom** [comment]
> Currently, I have no idea how I would access global names using a GDExtension instead of a module (outside the global names defined by the specific Language implementation itself), but maybe it exists, and I didn't find the right public api call for it.

You can use `ProjectSettings::get_singleton()->get_global_class_list()` to obtain the list of global classes.

The major con here is related to the queries that can be done through `ScriptServer`, like "is global class", "get global class base" and so on. Directly using `ScriptServer` means that it will perform the verification on already existing data. When using `ProjectSettings` we have not only to obtain the data every time we need to query something (it might have changed since the last query), but also re-implement the functions that perform those queries.

That's basically how I'm obtaining the information in my extension. *If* you decide to look into my public repositories (GDExtensionPack) and want to see it in practice, check the `internal.h` and `internal.cpp` files.

**Comment by dsnopek** [comment]
I think the next steps on this issue are:

1. Make a list of what functions from we want to expose from `ScriptServer` and get consensus among the folks responsible for the scripting API
2. Make a Godot PR adding functions to `gdextension_interface.h` for those
3. Make a godot-cpp PR adding a `ScriptServer` class that mimicks Godot's `ScriptServer` and calls those functions from `gdextension_interface.h`

**Comment by CedNaru** [comment]
I think the following methods are good candidates:
- `is_global_class`
- `get_global_class_path`
- `get_global_class_list`
- `get_global_class_base`
- `get_global_class_native_base`

Basically all methods that query data related to global classes. But no method that change the state of the ScriptServer should be exposed.

**Comment by Kehom** [comment]
After some thoughts, there is one aspect to consider. From my understanding the goal is that godot-cpp reaches (feature/capabilities) parity with modules. If just a subset of `ScriptServer` gets exposed to godot-cpp then this goal will not be reached. Moreover, I do believe there are several other internal singletons that are not exposed to GDExtension that could be easily used by modules, further "contributing" to a list of missing "features".

**Comment by dsnopek** [comment]
> After some thoughts, there is one aspect to consider. From my understanding the goal is that godot-cpp reaches (feature/capabilities) parity with modules. If just a subset of `ScriptServer` gets exposed to godot-cpp then this goal will not be reached.

While we do have the goal of feature parity between modules and GDExtension, I'm not sure we can ever reach that 100%. There will always be APIs that need to be public in the engine so that other parts of the engine can use them, but where we don't actually want modules to use them. Unfortunately, before GDExtension we didn't really do a good job of deciding or documenting exactly which those are. :-)

Since any API we expose via GDExtension becomes something we need to maintain compatibility going forward, this is a good time to try and decide on what those should be. And maybe that is something we can feed back into the engine, and document that modules also shouldn't use certain methods? We most often modify godot-cpp to match Godot, but another way to create API parity between them is to modify Godot to match godot-cpp :-)
